src/memory.h

```diff
@@ -172,4 +177,39 @@ class AllocatedMemory : public MutableMemory {
   ~AllocatedMemory() override;
 };
 
+#ifdef TRITON_ENABLE_GPU
+class GrowableMemory : public MutableMemory {
+ public:
+  // Create a continuous data buffer with 'byte_size', 'memory_type' and
+  // 'memory_type_id'. Note that the buffer may be created on different memory
+  // type and memory type id if the original request type and id can not be
+  // satisfied, thus the function caller should always check the actual memory
+  // type and memory type id before use.
+  GrowableMemory(
+      size_t byte_size, TRITONSERVER_MemoryType memory_type,
+      int64_t memory_type_id, std::unique_ptr<Allocation>&& allocation,
+      size_t virtual_page_size);
+
+  static Status Create(
+      std::unique_ptr<GrowableMemory>& growable_memory, size_t byte_size,
+      TRITONSERVER_MemoryType memory_type, int64_t memory_type_id,
+      size_t virtual_page_size);
+
+  Status Map(CUmemGenericAllocationHandle& block);
+
+  ~GrowableMemory() override;
+
+  Status Resize(size_t size);
+
+  std::unique_ptr<Allocation>& GetAllocation();
+
+ private:
+  std::unique_ptr<Allocation> allocation_;
```

Comment:
Does allocation need to be a unique ptr? could it be a instance variable instead - and initialized in the constructor?
from my understanding the allocation is empty to begin with and just has the device id?

Reply:
will be part of future changes for Allocation object.